Finalise application process

kdmnk opened the pull request:
This PR implements the missing finalise applications feature. 

It's on a new page now where the final names and statistics can be seen.
It is also possible to change who has been admitted as resident or extern.

kdmnk committed:
finalise application backend

kdmnk committed:
UI

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in dbf0da7 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/618

kdmnk committed:
fixes

kdmnk committed:
fix missing docs

kdmnk committed:
Merge branch 'development' into finalise-applications

viktorcsimma commented on app/Http/Controllers/Auth/AdmissionController.php:
```diff
@@ -165,54 +168,82 @@ public function updateNote(Request $request, Application $application): Redirect
+    public function indexFinalize(): View
+    {
+        $this->authorize('finalize', Application::class);
+        $admitted = Application::query()->with(['user', 'applicationWorkshops'])->admitted()->get()->sortBy('user.name');
```
Isn't it more efficient to do the ordering before the `get()`, while still in the query? With an `orderBy`.

kdmnk replied:
Then I would need to join the users table... If 'with' has other semantics (which I guess) then we still have some overhead. I think it is fine as it is

viktorcsimma replied:
OK:)

viktorcsimma commented on app/Http/Controllers/Auth/AdmissionController.php:
```diff
@@ -165,54 +168,82 @@ public function updateNote(Request $request, Application $application): Redirect
+                Storage::delete($file->path);
+            }
+            $files->delete();
+            // soft deletes application, keep them for future reference
+            // (see https://github.com/EotvosCollegium/mars/issues/332#issuecomment-2014058021)
```
Does the college qualify as a 'higher education institute'? (AFAIK, only the university itself has legal personality, not even the faculties.) But I'm obviously not a lawyer, and I can live without purging them anyway:)

horcsinbalint replied:
The College itself is not a `felsőoktatási intézmény`, but it is rather a `szervezeti egység` of the university which is a `felsőoktatási intézmény`. So I think the College should comply with the `Nemzeti Felsőoktatási Törvény`. Where the GDPR is concerned the `adatkezelő` is the university and the `felelős szervezeti egység` is the College AFAIK.

viktorcsimma replied:
I see... I'm not confident in these things; let this be so then.

viktorcsimma commented on app/Http/Controllers/Auth/AdmissionController.php:
```diff
@@ -165,54 +168,82 @@ public function updateNote(Request $request, Application $application): Redirect
+                $application->user->setStatusFor($this->semester(), SemesterStatus::ACTIVE);
+                $application->user->internetAccess->extendInternetAccess($this->semester()->getStartDate()->addMonth());
+            }
+            // delete data for not admitted users
+            $files = File::query()
+                ->whereIn('application_id', $not_admitted_applications->pluck('id')) // application files
+                ->orWhereIn('user_id', $not_admitted_applications->pluck('user_id')); // profile pictures
+            foreach ($files->get() as $file) {
+                Storage::delete($file->path);
+            }
```
Are these also only soft-deleted? If applications are kept, the files belonging to them should be kept as well.

kdmnk replied:
I only delete the not admitted applications' files. The others just stay there.

viktorcsimma replied:
Oh, sorry; I've missed that.

viktorcsimma commented on app/Http/Controllers/Auth/AdmissionController.php:
```diff
@@ -165,54 +168,82 @@ public function updateNote(Request $request, Application $application): Redirect
+            Application::whereNotIn('id', $admitted_applications->pluck('id'))->forceDelete();
+            ApplicationWorkshop::query()->delete();
+
+            // Note: users with a not submitted applications will also be deleted
```
Nitpick: there is an unnecessary `a` there.

BertalanD commented on app/Http/Controllers/Auth/AdmissionController.php:
```diff
@@ -165,54 +168,82 @@ public function updateNote(Request $request, Application $application): Redirect
+                ->with('application')
+                ->whereIn('id', $not_admitted->pluck('user_id'))
+                //ignore users with any existing role
+                ->whereDoesntHave('roles')
```
FYI: At the moment, even newly registered applicants get the printer role by default, so I don't think this will end up deleting anybody as-is. (last year's applicants are still present in the `users` table too)

https://github.com/EotvosCollegium/mars/blob/fabec6fb25c18d1c98a004e438094e62f5546fb2/app/Http/Controllers/Auth/RegisterController.php#L115

Probably easier to jump into the database and fix this manually, rather than changing the code here to handle this quirk.

kdmnk replied:
Huh, true... But this won't be the case after https://github.com/EotvosCollegium/mars/pull/394 has been merged. 
But even before that we can just delete that role.

viktorcsimma replied:
I agree; we don't need the printer role. No one can print without paying for admins before, anyway.

BertalanD commented on app/Http/Controllers/Auth/AdmissionController.php:
```diff
@@ -165,54 +168,82 @@ public function updateNote(Request $request, Application $application): Redirect
+        return view('auth.admission.finalize', [
+            'semester' => $this->semester(),
+            'admitted_applications' => $admitted,
+            'users_to_delete' => User::query()
+                ->withoutGlobalScope('verified')
+                ->with('application')
+                ->whereIn('id', $not_admitted->pluck('user_id'))
+                //ignore users with any existing role
+                ->whereDoesntHave('roles')
+                ->orderBy('name')
+                ->get()
```
Can we factor out this query and the identical one in `finalize` into a helper function?

Also, this looks prone to race conditions; what happens if someone messes with admission statuses between this page is rendered and the finalize button is pressed?

kdmnk replied:
Good point. But in real life I expect that the workshop committees will have a deadline to finalise the statuses. Or everything will be handled by the secretary/sys admins. 
Let's just pay attention to this as I don't have an easy solution now.

BertalanD replied:
> Let's just pay attention to this as I don't have an easy solution now.

Sounds good to me. But can you still do something with the code duplication for the query?

BertalanD commented on app/Http/Controllers/Auth/AdmissionController.php:
```diff
@@ -165,54 +168,82 @@ public function updateNote(Request $request, Application $application): Redirect
+            }
+            $files->delete();
+            // soft deletes application, keep them for future reference
+            // (see https://github.com/EotvosCollegium/mars/issues/332#issuecomment-2014058021)
```
@horcsinbalint

horcsinbalint replied:
I have not checked thoroughly, but this part LGTM. If it deletes all the data for rejected users but keeps the data for admitted users, it should be fine AFAIK (I think it is the behavior that we should aim to achieve).

One question is what happens in cases where a student starts a `Jogorvoslati eljárás`. I would not want to go into details publicly about our backup strategy (as it is proprietary) but I think that should solve that issue.

horcsinbalint replied:
But the question about `Jogorvoslati eljárás` is not a real issue IMO.

viktorcsimma commented on resources/views/auth/admission/index.blade.php:
```diff
@@ -73,27 +73,9 @@
     <hr>
     <h6>Összesen: <b class="right">{{$applications->count()}} jelentkező</b></h6>
     @can('finalize', \App\Models\Application::class)
-        @if($applicationDeadline?->addWeeks(1) < now())
```
index.blade.php:53 adds a week to $applicationDeadline without making a copy of it. It works for now because it isn't used at the end, but maybe it would be more future-proof.

viktorcsimma commented on resources/views/auth/admission/finalize.blade.php:
```diff
@@ -0,0 +1,80 @@
+            <div class="row">
+                <div class="col">
+                    Hogyha a felvételi eljárás befejeződött, akkor a felvett jelentkezőket itt lehet
+                    jóváhagyni. A felvettek státusza a {{ $semester->tag }} szemeszterre automatikusan aktív lesz.
```
This throws an exception if the periodic event is not yet set.

Also, for me, this is the spring semester and not the autumn one as it should be.

kdmnk replied:
The semester depends on the periodic event. You should set it for the autumn one. In the live system it looks fine.

viktorcsimma replied:
Oh, I see. Sorry then.

viktorcsimma commented on resources/views/auth/admission/finalize.blade.php:
```diff
@@ -0,0 +1,80 @@
+                        </a><br/>
+                    @endforeach
+                </div>
+                <div class="col">
+                    <blockquote>A művelet végrehajtása előtt ajánlott biztonsági mentést készíteni az adatbázisról.</blockquote>
+                    <form id="finalize-application-process" method="POST"
+                          action="{{route('admission.finalize')}}">
+                        @csrf
+                        <x-input.button class="red" text="A felvételi lezárása"/>
+                    </form>
+                </div>
```
This should not appear if we have not reached `$applicationDeadline->copy()->addWeek()` yet.

viktorcsimma commented on app/Http/Controllers/Auth/AdmissionController.php:
```diff
@@ -165,54 +168,82 @@ public function updateNote(Request $request, Application $application): Redirect
-        //        Cache::clear();
-        // return back()->with('message', 'Sikeresen jóváhagyta az elfogadott jelentkezőket');
-        return back()->with('error', 'Még nincs implementálva.');
+        $this->authorize('finalize', Application::class);
```
Here, we should check whether we have reached `$applicationDeadline->copy()->addWeek()`.

Or should we abolish this requirement? (I think it is useful, but I do not insist on it.) If not more, I think we should not allow finalisation while the application period is open. (The finalisation page could be shown, that is useful; but without the option to actually finalise applications.)

kdmnk replied:
I wanted to delete that part (maybe I left some parts of this logic). Originally, i just wanted to avoid accidental finalisations but now we have a separate page only for that, I think it is clear now.

BertalanD replied:
FWIW, removing that sounds okay to me (the one-week delay struck me as arbitrary when I first saw it)

viktorcsimma replied:
I think it would still be useful to prevent finalisation while the application period is open. But I can live without that; just remove the remainders and warnings related to that logic then.

kdmnk replied:
I have put it back but it only checks for the application deadline, does not add a week.

viktorcsimma replied:
Nice:)

viktorcsimma reviewed (changes requested):
Also, a seeded applicant did not get workshops nor educational information; is this specific to seeded users or a general problem? I think at least workshops should not be a problem.

BertalanD commented on app/Livewire/ApplicationRoleStatusUpdate.php:
```diff
@@ -0,0 +1,53 @@
+     * Update whether the applicant has been admitted as resident or not.
+     * @param $workshop
+     */
+    public function switchResidentRole()
+    {
+        $this->application->update(['admitted_for_resident_status' => !$this->application->admitted_for_resident_status]);
+        $this->lastUpdated = Carbon::now();
+    }
```
Does this need access control?

BertalanD commented:
@coderabbitai review

kdmnk committed:
fix PR comments

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in f33f2f2 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/618

kdmnk committed:
Merge branch 'development' into finalise-applications

kdmnk committed:
fix tests

viktorcsimma commented on resources/views/auth/admission/index.blade.php:
```diff
@@ -73,26 +72,10 @@
-                    </div>
-                </div>
-            </div>
+        @if($applicationDeadline && $applicationDeadline < now()))
```
Here, there is an extra `)` which appears on the generated page.

kdmnk replied:
I don't have my laptop for the rest of the weekend, if you want to merge could you fix it?

viktorcsimma reviewed (changes requested):
LGTM now:) one more small change and we can go ahead, I think.

viktorcsimma commented:
@coderabbitai review

kdmnk committed:
fix typo

kdmnk committed:
Merge branch 'development' into finalise-applications

kdmnk committed:
fix wrong import
